flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/StopWithSavepoint.java

```diff
@@ -0,0 +1,214 @@
+    private final CompletableFuture<String> operationCompletionFuture;
+    private final boolean advanceToEndOfEventTime;
+    private final String targetDirectory;
+    private final CheckpointCoordinator checkpointCoordinator;
+    private final Context context;
+
+    StopWithSavepoint(
+            Context context,
+            ExecutionGraph executionGraph,
+            ExecutionGraphHandler executionGraphHandler,
+            OperatorCoordinatorHandler operatorCoordinatorHandler,
+            Logger logger,
+            ClassLoader userCodeClassLoader,
+            CheckpointCoordinator checkpointCoordinator,
+            String targetDirectory,
+            boolean advanceToEndOfEventTime,
+            CompletableFuture<String> operationCompletionFuture) {
+        super(
+                context,
+                executionGraph,
+                executionGraphHandler,
+                operatorCoordinatorHandler,
+                logger,
+                userCodeClassLoader);
+        this.context = context;
+        this.checkpointCoordinator = checkpointCoordinator;
+        this.targetDirectory = targetDirectory;
+        this.advanceToEndOfEventTime = advanceToEndOfEventTime;
+        this.operationCompletionFuture = operationCompletionFuture;
+    }
+
+    @Override
+    public void onEnter() {
+        final ExecutionGraph executionGraph = getExecutionGraph();
+        // we stop the checkpoint coordinator so that we are guaranteed
+        // to have only the data of the synchronous savepoint committed.
+        // in case of failure, and if the job restarts, the coordinator
+        // will be restarted by the CheckpointCoordinatorDeActivator.
+        checkpointCoordinator.stopCheckpointScheduler();
+
+        CompletableFuture<String> savepointFuture =
+                checkpointCoordinator
+                        .triggerSynchronousSavepoint(advanceToEndOfEventTime, targetDirectory)
+                        // handle errors while creating the savepoint
+                        .handleAsync(
+                                this::handleSavepointOperationResult,
+                                context.getMainThreadExecutor())
+                        // Wait for job to be finished
+                        .thenCompose(
+                                path ->
+                                        executionGraph
+                                                .getTerminationFuture()
+                                                .thenApply(
+                                                        status ->
+                                                                new JobStatusAndSavepointPath(
+                                                                        status, path)))
+                        // check that the job finished successfully
+                        .handleAsync(
+                                this::handleStopWithSavepointOperationResult,
```

Comment:
Why are we handling the final job result by accessing the `getTerminationFuture`? Don't we already have `onGloballyTerminalState` which tells us if the `ExecutionGraph` has reached a globally terminal state?